core/src/main/java/com/graphhopper/routing/profiles/UnsignedIntEncodedValue.java

```diff
@@ -171,12 +172,76 @@ public boolean equals(Object o) {
     }
 
     @Override
-    public int hashCode() {
-        return Objects.hash(name, fwdDataIndex, bwdDataIndex, bits, maxValue, fwdShift, bwdShift, fwdMask, bwdMask, storeTwoDirections);
+    public final int hashCode() {
+        return getVersion();
     }
 
     @Override
     public int getVersion() {
-        return hashCode();
+        int val = staticHashCode(name);
+        val = 31 * val + (storeTwoDirections ? 1231 : 1237);
+        return staticHashCode(val, fwdDataIndex, bwdDataIndex, bits, maxValue, fwdShift, bwdShift, fwdMask, bwdMask);
+    }
+
+    /**
+     * Produces a static hashcode for an integer arrays that is platform independent and still compatible to the default
+     * of openjdk.
+     *
+     * @see Arrays#hashCode(int[])
+     */
+    static int staticHashCode(int... vals) {
+        if (vals == null)
+            return 0;
+        int len = vals.length;
+        int[] locVal = vals;
+        int val = 1;
+        for (int idx = 0; idx < len; ++idx) {
+            val = 31 * val + locVal[idx];
+        }
+
+        return val;
+    }
+
+    /**
+     * Produces a static hashcode for an Enum arrays that is platform independent and still compatible to the default
+     * of openjdk.
+     */
+    static int staticHashCode(Enum... vals) {
+        if (vals == null)
+            return 0;
+        int len = vals.length;
+        Enum[] locVal = vals;
```

Comment:
very minor, but do we really need the extra `locVal` reference or can we not just do `vals[idx]` in the loop below ?

Reply:
I read somewhere (cannot find the link now) that a local variable can have different behaviour (faster). When thinking where this can come for one explanation would make sense if you assign an instance variable to a local variable (moving it from heap to stack). But here it is basically a method parameter (stack) to stack. There must be a reason as this is done e.g. in `Arrays#hashCode`. But obviously the stack allocation has to be larger and also more bytecode lines are produced (checked via `javap -v`) and [some say they are identical](https://coderanch.com/t/556166/java/Parameter-Variable-considered-Local-Variable). Will remove it and read a bit more about the JVM internals.

Reply:
> Will remove it and read a bit more about the JVM internals.

Cool looking forward to hear what you find out.

Reply:
Haven't found much. Just tested it in JMH and it looks that between arrays as parameter or local variable there is no difference, but as a class variable the code is indeed ~5% slower:

https://gist.github.com/karussell/4cd7999235c651f7e714630b61297c6b